system/lib/wasmfs/wasmfs.cpp

```diff
@@ -8,60 +8,258 @@
+  time_t atime;  // Time when the content was last accessed
+
+  std::vector<std::shared_ptr<block_t>> blocks; // pointers to data blocks
+
+  std::mutex file_mutex;
+
+public:
+  // TODO: stubbed virtual function - convert to pure virtual function later
+  virtual __wasi_errno_t read(const __wasi_iovec_t* iovs, size_t iovs_len, __wasi_size_t* nread) {
+    return 1;
+  };
+  virtual __wasi_errno_t write(
+    const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) = 0;
+};
+
+class StdinFile : public File {
+public:
+  // TODO: fill in write for stdin
+  __wasi_errno_t write(
+    const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) override {
+    return 0;
+  }
+};
+
+__wasi_errno_t write_buffer(const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten,
+  void (*func)(const char*)) {
+  __wasi_size_t num = 0;
+  for (size_t i = 0; i < iovs_len; i++) {
+    const uint8_t* buf = iovs[i].buf;
+    __wasi_size_t len = iovs[i].buf_len;
+    for (__wasi_size_t j = 0; j < len; j++) {
+      uint8_t current = buf[j];
+      if (current == 0 || current == 10) {
+        fd_write_stdstream_buffer.push_back('\0'); // for null-terminated C strings
+        func(&fd_write_stdstream_buffer[0]);
+        fd_write_stdstream_buffer.clear();
+      } else {
+        fd_write_stdstream_buffer.push_back(current);
+      }
+    }
+    num += len;
+  }
+  *nwritten = num;
+  return 0;
+}
+
+class StdoutFile : public File {
+public:
+  __wasi_errno_t write(
+    const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) override {
+    return write_buffer(iovs, iovs_len, nwritten, &emscripten_console_log);
+  }
+};
+
+class StderrFile : public File {
+public:
   // TODO: May not want to proxy stderr (fd == 2) to the main thread.
   // This will not show in HTML - a console.warn in a worker is suffficient.
   // This would be a change from the current FS.
-  if (fd == 1 || fd == 2) {
-    __wasi_size_t num = 0;
-    for (size_t i = 0; i < iovs_len; i++) {
-      const uint8_t* buf = iovs[i].buf;
-      __wasi_size_t len = iovs[i].buf_len;
-      for (__wasi_size_t j = 0; j < len; j++) {
-        uint8_t current = buf[j];
-        if (current == 0 || current == 10) {
-          fd_write_stdstream_buffer.push_back('\0'); // for null-terminated C strings
-          if (fd == 1) {
-            emscripten_console_log(&fd_write_stdstream_buffer[0]);
-          } else if (fd == 2) {
-            emscripten_console_error(&fd_write_stdstream_buffer[0]);
-          }
-          fd_write_stdstream_buffer.clear();
-        } else {
-          fd_write_stdstream_buffer.push_back(current);
-        }
+  __wasi_errno_t write(
+    const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) override {
+    return write_buffer(iovs, iovs_len, nwritten, &emscripten_console_error);
+  }
+};
+
+class OpenFile {
+  std::shared_ptr<File> file_node;
+  uint32_t _offset;
```

Comment:
This should probably be `off_t` (whatever that is defined as) rather than `uint32_t` to match the types in the `lseek` interface.